Quantum-Control-Applications/Superconducting/Single Fixed Transmon/readout_duration_opt.py

```diff
@@ -0,0 +1,221 @@
+    plt.plot(arr.imag)
+    plt.show()
+
+
+def subtract_complex_arrays(arr1, arr2):
+    return arr1 - arr2
+
+
+def normalize_complex_array(arr):
+    # Calculate the simple norm of the complex array
+    norm = np.sqrt(np.sum(np.abs(arr) ** 2))
+
+    # Normalize the complex array by dividing it by the norm
+    normalized_arr = arr / norm
+
+    # Rescale the normalized array so that the maximum value is 1
+    max_val = np.max(np.abs(normalized_arr))
+    rescaled_arr = normalized_arr / max_val
+
+    return rescaled_arr
+
+
+def plot_three_complex_arrays(arr1, arr2, arr3):
+    fig, (ax1, ax2, ax3) = plt.subplots(1, 3, figsize=(15, 5))
+    ax1.plot(arr1.real)
+    ax1.plot(arr1.imag)
+    ax1.set_title("ground state")
+    ax2.plot(arr2.real)
+    ax2.plot(arr2.imag)
+    ax2.set_title("excited state")
+    ax3.plot(arr3.real)
+    ax3.plot(arr3.imag)
+    ax3.set_title("optimal weights")
+    plt.show()
+
+
+###################
+# The QUA program #
+###################
+
+number_of_divisions = 250  # number of chunks to divide the integration weights into // put something meaningful here
+division_length = int(readout_len / (4 * number_of_divisions))  # in clock cycles
+print("Integration weights chunk-size length in clock cyclces:", division_length)
+
+n_avg = 1e4  # number of averages
+cooldown_time = 5 * qubit_T1 // 4  # thermal decay time of the qubit
+
+with program() as opt_weights:
+    n = declare(int)
+    II = declare(fixed, size=number_of_divisions)
+    IQ = declare(fixed, size=number_of_divisions)
+    QI = declare(fixed, size=number_of_divisions)
+    QQ = declare(fixed, size=number_of_divisions)
+    ind = declare(int)
+
+    n_st = declare_stream()
+    II_st = declare_stream()
+    IQ_st = declare_stream()
+    QI_st = declare_stream()
+    QQ_st = declare_stream()
+
+    with for_(n, 0, n < n_avg, n + 1):
+        # ground state
+        measure(
+            "readout",
+            "resonator",
+            None,
+            demod.accumulated("cos", II, division_length, "out1"),
+            demod.accumulated("sin", IQ, division_length, "out2"),
+            demod.accumulated("minus_sin", QI, division_length, "out1"),
+            demod.accumulated("cos", QQ, division_length, "out2"),
+        )
+
+        with for_(ind, 0, ind < number_of_divisions, ind + 1):
+            save(II[ind], II_st)
+            save(IQ[ind], IQ_st)
+            save(QI[ind], QI_st)
+            save(QQ[ind], QQ_st)
+
+        wait(cooldown_time, "resonator")
+
+        align()
+
+        # excited state
+        play("x180", "qubit")
+        align("qubit", "resonator")
+        measure(
+            "readout",
+            "resonator",
+            None,
+            demod.accumulated("cos", II, division_length, "out1"),
+            demod.accumulated("sin", IQ, division_length, "out2"),
+            demod.accumulated("minus_sin", QI, division_length, "out1"),
+            demod.accumulated("cos", QQ, division_length, "out2"),
```

Comment:
same here?

Reply:
I don't think so we want to see how much SNR we have for diff readout lenghts

Reply:
Right sorry about that...